File: src/com/alchitry/labs/parsers/tools/lucid/ArrayBounds.java
package com.alchitry.labs.parsers.tools.lucid;

public class ArrayBounds {
	private int min, max;

	public ArrayBounds() {

	}

	public ArrayBounds(int i, int j) {
		setValues(i, j);
	}

	public void setValues(int i, int j) {
		max = i > j ? i : j;
		min = i > j ? j : i;
	}

	public boolean fitInWidth(int w) {
		if (max < w && min >= 0)
			return true;
		return false;
	}

	public boolean testBounds(ArrayBounds ab) {
		if (max < ab.getMax())
			return false;
		if (min > ab.getMin())
			return false;
		return true;
	}

	public boolean testBounds(int i, int j) {
		int bmax = i > j ? i : j;
		int bmin = i > j ? j : i;
		if (max < bmax)
			return false;
		if (min > bmin)
			return false;
		return true;
	}

	public boolean testBounds(int i) {
		if (i > max || i < min)
			return false;
		return true;
	}

	public int getMax() {
		return max;
	}

	public int getMin() {
		return min;
	}

	public int getWidth() {
		return max - min + 1;
	}

	@Override
	public String toString() {
		return "[" + max + ", " + min + "]";
	}
}

File: src/com/alchitry/labs/parsers/tools/lucid/ArrayBounds.kt
package com.alchitry.labs.parsers.tools.lucid

import java.lang.Integer.max
import java.lang.Integer.min

class ArrayBounds() {
    var min = 0
        private set
    var max = 0
        private set

    constructor(i: Int, j: Int) : this() {
        setValues(i, j)
    }

    fun setValues(i: Int, j: Int) {
        max = max(i, j)
        min = min(i, j)
    }

    fun fitInWidth(w: Int): Boolean {
        return max < w && min >= 0
    }

    fun testBounds(ab: ArrayBounds): Boolean {
        if (max < ab.max) return false
        return min <= ab.min
    }

    fun testBounds(i: Int, j: Int): Boolean {
        val bmax = if (i > j) i else j
        val bmin = if (i > j) j else i
        if (max < bmax) return false
        return min <= bmin
    }

    fun testBounds(i: Int): Boolean {
        return !(i > max || i < min)
    }

    val width: Int
        get() = max - min + 1

    override fun toString(): String {
        return "[$max, $min]"
    }
}